pkg/schedulingtypes/jobscheduler.go

```diff
@@ -0,0 +1,229 @@
+	return j.fedClient.SchedulingV1alpha1().JobSchedulingPreferences(j.controllerConfig.TargetNamespace).Watch(options)
+}
+
+func (j *JobScheduler) Start(stopChan <-chan struct{}) {
+	// NOOP
+}
+
+func (j *JobScheduler) HasSynced() bool {
+	return j.plugin.HasSynced()
+}
+
+func (j *JobScheduler) Stop() {
+	j.plugin.Stop()
+}
+
+func (j *JobScheduler) Reconcile(obj pkgruntime.Object, qualifiedName QualifiedName) ReconciliationStatus {
+	jsp, ok := obj.(*fedschedulingv1a1.JobSchedulingPreference)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Incorrect runtime object for JSP: %v", jsp))
+		return StatusError
+	}
+
+	clusterNames, err := j.plugin.ReadyClusterNames()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to get cluster list: %v", err))
+		return StatusError
+	}
+	if len(clusterNames) == 0 {
+		// no joined clusters, nothing to do
+		return StatusAllOK
+	}
+
+	result := j.schedule(jsp, qualifiedName, clusterNames)
+	err = j.ReconcileFederationTargets(j.fedClient, qualifiedName, result)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to reconcile Federation Targets for JSP named %s: %v", qualifiedName, err))
+		return StatusError
+	}
+
+	return StatusAllOK
+}
+
+func (j *JobScheduler) ReconcileFederationTargets(fedClient fedclientset.Interface, qualifiedName QualifiedName, result *JobScheduleResult) error {
+	err := j.plugin.ReconcilePlacement(qualifiedName, result)
+	if err != nil {
+		return err
+	}
+
+	err = j.plugin.ReconcileOverride(qualifiedName, result)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type ClusterJobValues struct {
+	Parallelism int32
+	Completions int32
+}
+
+type JobScheduleResult struct {
+	Result map[string]ClusterJobValues
+}
+
+func (r *JobScheduleResult) ClusterNames() []string {
+	clusterNames := []string{}
+	for name := range r.Result {
+		clusterNames = append(clusterNames, name)
+	}
+	return clusterNames
+}
+
+func (r *JobScheduleResult) SetPlacementSpec(obj *unstructured.Unstructured) {
+	obj.Object[util.SpecField] = map[string]interface{}{
+		util.ClusterNamesField: r.ClusterNames(),
+	}
+}
+
+// TODO (irfanurrehman): in this pass we ignore updates for jobs
+func (r *JobScheduleResult) PlacementUpdateNeeded(names []string) bool {
+	return false
+}
+
+func (r *JobScheduleResult) SetOverrideSpec(obj *unstructured.Unstructured) {
+	overrides := []interface{}{}
+	for clusterName, values := range r.Result {
+		overridesMap := map[string]interface{}{
+			util.ClusterNameField: clusterName,
+			ParallelismField:      values.Parallelism,
+			CompletionsField:      values.Completions,
+		}
+		overrides = append(overrides, overridesMap)
+	}
+	obj.Object[util.SpecField] = map[string]interface{}{
+		util.OverridesField: overrides,
+	}
+}
+
+// TODO (irfanurrehman): in this pass we ignore updates for jobs
+func (r *JobScheduleResult) OverrideUpdateNeeded(typeConfig typeconfig.Interface, obj *unstructured.Unstructured) bool {
```

Comment:
Given that this method appears to have been implemented in the fake method for the test, and why should this method remain unimplemented in this PR?